File: LeetCode/Problem 56.py
# LeetCode: Problem 56
# Jonathan Kosasih

#Given a collection of intervals, merge all overlapping intervals.
#
#Example 1:
#Input: [[1,3],[2,6],[8,10],[15,18]]
#Output: [[1,6],[8,10],[15,18]]
#Explanation: Since intervals [1,3] and [2,6] overlaps, merge them into [1,6].
#
#Example 2:
#Input: [[1,4],[4,5]]
#Output: [[1,5]]
#Explanation: Intervals [1,4] and [4,5] are considered overlapping.
#------------------------------------------------------------------
class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        if(not intervals):
            return []
        intervals = sorted(intervals, key=lambda entry: entry[0])
        ans = [intervals[0]]
        for index in range(1,len(intervals)):
            if(ans[-1][-1] >= intervals[index][0]):
                ans[-1][-1] = max(ans[-1][-1], intervals[index][1])
            else:
                ans.append(intervals[index])
        return ans
        
sol = Solution()
example = [[1,3],[2,6],[8,10],[15,18]]
print(sol.merge(example))
example = [[1,4],[4,5]]
print(sol.merge(example))
example = [[1,4],[0,4]]
print(sol.merge(example))
example = [[1,4],[0,1]]
print(sol.merge(example))
example = [[1,4],[0,0]]
print(sol.merge(example))

File: LeetCode/C++/Problem 56.cpp
// LeetCode: Problem 56
// Jonathan Kosasih
/*
Given an array of non-negative integers, you are initially positioned at the first index of the array.

Each element in the array represents your maximum jump length at that position.

Determine if you are able to reach the last index.
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;

class Compare {
public:
	bool operator () (vector<int>& vec1, vector<int>& vec2) {
		return vec1[0] < vec2[0];
	}
};

Compare CompObject;

class Solution {
public:
	vector<vector<int>> merge(vector<vector<int>>& intervals) {
		if (intervals.empty())
			return {};

		sort(intervals.begin(), intervals.end(), CompObject);

		int i = 0;
		vector<vector<int>> res;
		while (i < intervals.size() - 1) {
			if (intervals[i][1] < intervals[i + 1][0])
				res.push_back(intervals[i]);

			else if (intervals[i][1] >= intervals[i + 1][0] && intervals[i][1] < intervals[i + 1][1])
				intervals[i + 1][0] = intervals[i][0];

			else {
				intervals[i + 1][0] = intervals[i][0];
				intervals[i + 1][1] = intervals[i][1];
			}
			i++;
		}

		res.push_back(intervals[i]);
		return res;
	}
};

int main()
{
    //std::cout << "Hello World!\n";
}